server/src/main/java/org/elasticsearch/plugins/PluginIntrospector.java

```diff
@@ -67,36 +79,66 @@ static PluginIntrospector getInstance() {
+    }
+
+    /**
+     * Returns the list of deprecated Elasticsearch plugin interfaces which are implemented by the
+     * given plugin implementation class. The list contains the simple names of the interfaces.
+     */
+    List<String> deprecatedInterfaces(final Class<?> pluginClass) {
+        assert Plugin.class.isAssignableFrom(pluginClass);
+        return interfaceClasses(pluginClass, deprecatedPluginClasses::contains).map(Class::getSimpleName).sorted().toList();
+    }
+
+    /**
+     * Returns the deprecated methods from Elasticsearch plugin interfaces which are implemented by
+     * the given plugin implementation class. The map is from the simple method name to the simple interface class name.
+     *
+     * @apiNote The simple names work as a key because they are unique across all plugin interfaces.
+     */
+    Map<String, String> deprecatedMethods(final Class<?> pluginClass) {
+        return findOverriddenMethods(pluginClass, pluginDeprecatedMethodsMap);
+    }
+
+    // finds the subset of given methods that are overridden by the given class
+    // returns a map of method name to interface name the method was declared in
+    private Map<String, String> findOverriddenMethods(final Class<?> pluginClass, Map<Class<?>, List<MethodType>> methodsMap) {
         assert Plugin.class.isAssignableFrom(pluginClass);
-        List<Class<?>> esPluginClasses = Stream.concat(Stream.of(Plugin.class), interfaceClasses(pluginClass)).toList();
+        List<Class<?>> clazzes = Stream.concat(Stream.of(Plugin.class), interfaceClasses(pluginClass, methodsMap::containsKey)).toList();
+        if (clazzes.isEmpty()) {
+            return Map.of();
+        }
 
-        List<String> overriddenMethods = new ArrayList<>();
-        for (var esPluginClass : esPluginClasses) {
-            List<MethodType> esPluginMethods = pluginMethodsMap.get(esPluginClass);
-            assert esPluginMethods != null : "no plugin methods for " + esPluginClass;
-            for (var mt : esPluginMethods) {
+        Map<String, String> overriddenMethods = new HashMap<>();
+        for (var clazz : clazzes) {
+            List<MethodType> methods = methodsMap.get(clazz);
+            if (methods == null) {
+                continue;
+            }
+            for (var mt : methods) {
                 try {
                     Method m = pluginClass.getMethod(mt.name(), mt.parameterTypes());
-                    if (m.getDeclaringClass() == esPluginClass) {
+                    if (m.getDeclaringClass() == clazz) {
                         // it's not overridden
                     } else {
-                        assert esPluginClass.isAssignableFrom(m.getDeclaringClass());
-                        overriddenMethods.add(mt.name());
+                        assert clazz.isAssignableFrom(m.getDeclaringClass());
+                        overriddenMethods.put(mt.name(), clazz.getSimpleName());
```

Comment:
We should assert that overridenMethods.get(mt.name()) == null. Just in case someone mistakenly duplicates a name in two different interfaces.